### ISSUE TITLE
Large file transfers (from s_server) fails if built with Kernel TLS enabled. 

### ISSUE THREAD

**Issue by Griffin86** [openssl/openssl#8491]
While testing the (relatively new) KTLS TX path functionality (with my own build of OpenSSL from the master branch - further details below) using s_client and s_server, I observed that file transfers larger than 44 KB fail to complete.

Looking at PCAPs of a successful transfer (44 KB) vs a failed transfer (2880 KB) (attached along with the server RSA key to allow for decryption in Wireshark), the following is worth noting:

1. The TLS hanshake completes successfully in both cases.
2. Purely looking at the PCAP of the failed transfer, it seems as if the server simply stops sending records to the client after a few records have been transferred successfully (i.e. the client still responds with a TCP ACK after receiving the 'last successful record', but the server does not continue sending any more data).
3. For the failed transfer, the TCP ACK FIN sent by the client (at the end of the capture) was manually initiated with Ctrl-C on the client terminal. (Once the server stops sending packets, the connection remains open, but the client appears to 'hang'. Even after several minutes the client does not time out - hence the need to close the connection manually).

The example files I used to determine this are attached. Please note that while the files simply contain garbage data, a 'tag' was added at the end (#thisisendoffile) to simplify debugging (i.e. to quickly be able to determine if the whole file was transferred or not). 

Additional observances are noted at the end of this report (after mentioning all version/device related information). 

[KTLS_TX_bug_report.tar.gz](https://github.com/openssl/openssl/files/2970888/KTLS_TX_bug_report.tar.gz)

Note: I haven't determined the exact file size 'failure point', but I am consistently able to successfully transfer 44 KB files. 

Command used to start s_server:
```
/root/dstreicher/openssl_ktls_isolated/prefix_dir/bin/openssl s_server -accept 443 -tls1_2 -WWW -cert /root/dstreicher/nginx_custom_config_deployment/certs/rsacert_2048.pem -key /root/dstreicher/nginx_custom_config_deployment/certs/rsakey_2048.pem
```
Command used to request file from client:
```
echo -e "GET /target_filename \r\n\r\n" | /root/dstreicher/openssl_ktls_isolated/prefix_dir/bin/openssl s_client -quiet -connect 10.9.0.100:443 -cipher 'AES128-GCM-SHA256' -tls1_2
```
Note: The AES128-GCM-SHA256 cipher is used to allow for decryption of TLS data when analysing the PCAP files in Wireshark.

OpenSSL version:
```
Master branch (Commit ID: c37e635065a5bcb744dfb282f59e978e41490ce4)
```
OpenSSL build options (configure command):
```
./config --prefix=/root/dstreicher/openssl_ktls_isolated/prefix_dir/ --openssldir=/root/dstreicher/openssl_ktls_isolated/openssl_dir/ enable-ktls zlib shared threads enable-weak-ssl-ciphers "-Wl,-rpath,'"'$$'"ORIGIN/../lib'"
```

Hardware setup used:
s_client (host 1) <===wire===> s_server (host 2)

Host 1 OS (kernel) information:
```
  Operating System: Ubuntu 18.10
            Kernel: Linux 4.19.0+
      Architecture: x86-64
```

Host 2 OS (kernel) information:
(Note: the issue was first seen with Kernel 4.19 on the server side as well. In an attempt to determine if the issue may be Kernel related, I upgraded the server side to Kernel 5.0. Unfortunately this did not resolve the issue.)
```
  Operating System: Ubuntu 18.10
            Kernel: Linux 5.0.0+
      Architecture: x86-64
```

Other important caveats/strange observances:

1. When building OpenSSL without KTLS support (i.e. using same build command but omitting the 'enable-ktls' flag) the file transfer issue does not occur.
2. When running both s_server and s_client from the same host, the issue does not occur. My guess at this point is that something may be amiss with the KTLS TX buffer handling (or error handling perhaps?), which only surfaces once the client and server are 'separated' by a physical network link).
3. The failures occur regardless of whether or not the client side is built with KTLS support or not.
4. When running s_server with the -debug flag, the larger file transfer succeeds. This is perhaps the strangest factor: my guess would be that with OpenSSL running at a slower speed the KTLS socket has more time to actually transmit the data to the wire. I may very well be completely wrong about this however.
5. If the 'echo' is removed from the s_client command as follows:
```
/root/dstreicher/openssl_ktls_isolated/prefix_dir/bin/openssl s_client -quiet -connect 10.9.0.100:443 -cipher 'AES128-GCM-SHA256' -tls1_2
```
and the GET request is 'manually' initiated after the TLS handshake is complete by typing 
```
 GET 2880KB.bin_name
```
the file transfer still 'halts' after a few records were transferred, but if the 'return' key is pressed (which I believe the server interprets as a HTTP continuation packet) the transfer continues for a while and then 'halts' again. This can be repeated until the entire file is transferred. (Once again, this behaviour does not occur if the server is built without KTLS support).
6. I have observed odd occurrences where the larger file transfer actually succeeds, but this case is not readily reproducible (and in the far minority).

Please let me know if I need to provide any additional information.

**Comment by mattcaswell** [comment]
Ping @borisPis

**Comment by Griffin86** [comment]
I reverted the kernel on my server (host 2) to v4.20, which seems to resolve the issue. 
Since this seems to actually be a Kernel bug, I think this bug report can be closed?

(In summary then: Kernel 4.19 contains some KTLS TX breakages, which are fixed in 4.20, but is broken again in Kernel 5.0 -  may possibly even be different TX issues in the Kernel itself).

**Comment by mattcaswell** [comment]
Seems to be a kernel bug so closing.
